### PR TITLE
GOBBLIN-687: Pass TopologySpec map to DagManager to allow reuse of Sp…

### PR DESCRIPTION
**Pull request by sv2000**
…ecExecutors during DAG deserialization.

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-687

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
DagManager maintains state of all currently executing DAGs, by serializing each DAG on compilation and persisting it to a durable store. The serialized DAG includes Job config as well as the SpecExecutor config for each job in the DAG. This is done to correctly resume execution of DAGs in case of service restarts or leadership change. 

Currently, on service restart/leadership change, the new master de-serializes SpecExecutor config and creates a SpecExecutor instance for each job in the DAG. If the number of DAGs is large, this can result in many connections to the underlying executor instance. The proposed fix allows the DagManager to re-use the SpecExecutor instances created by the TopologySpecFactory when it deserializes a DAG.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Enhanced unit tests in FSDagStateStoreTest and tested in local environment.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by sv2000** [comment]
@autumnust please review this PR.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java [hunk 1]
```diff
@@ -107,6 +104,10 @@ public Orchestrator(Config config, Optional<TopologyCatalog> topologyCatalog, Op
         | ClassNotFoundException e) {
       throw new RuntimeException(e);
     }
+    //Pass the topologySpecMap to the DagManager
```
Comment (autumnust) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java]: Shall we move this block under line 93?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java [hunk 2]
```diff
@@ -218,6 +211,12 @@ public GobblinServiceManager(String serviceName, String serviceId, Config config
       this.helixManager = Optional.absent();
     }
 
+    // Initialize DagManager
```
Comment (autumnust) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java]: Initially `DagManager` is only initialized when `isFlowCatalogEnabled` set to true, but moving it out means this condition doesn't apply anymore.  Is that on purpose ?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java [hunk 3]
```diff
@@ -354,7 +354,10 @@ private void handleLeadershipChange(NotificationContext changeContext) {
       }
 
       if (this.isDagManagerEnabled) {
-        this.dagManager.setActive(true);
+        //Activate DagManager only if TopologyCatalog is initialized. If not; skip activation.
+        if (this.topologyCatalog.getInitComplete().getCount() <= 0) {
```
[the remark this answers is not in the document]
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java]: Changed it to == 0.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java (hunk 1 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java]: We need this after the spec compiler initialization, to make sure that the topologySpecMap has been instantiated. Added comment to make it explicit.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java (hunk 2 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java]: Thanks for noticing this! The DagManager initialization does not depend on whether FlowCatalog is enabled or not. We have a separate config to enable/disable DagManager. I can't recall why it was inside if (isFlowCatalogEnabled). Yes, the move outside the if block is intentional.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java [hunk 4]
```diff
@@ -232,6 +231,7 @@ public GobblinServiceManager(String serviceName, String serviceId, Config config
       this.serviceLauncher.addService(this.scheduler);
     }
 
+
```
[the remark this answers is not in the document]
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java]: Done.

**Comment by sv2000** [comment]
@autumnust addressed your comments. Can you please take another look?

**Comment by autumnust** [comment]
@sv2000 LGTM.

**Pull request closed without merging**